monai/networks/nets/regunet.py

```diff
@@ -0,0 +1,249 @@
+            RegistrationResidualConvBlock(
+                spatial_dims=self.spatial_dims,
+                in_channels=out_channels,
+                out_channels=out_channels,
+                kernel_size=kernel_size,
+            ),
+        )
+
+    def build_decode_layers(self):
+        # decoding / up-sampling
+        # [depth - 1, depth - 2, ..., min_extract_level]
+        self.decode_deconvs = nn.ModuleList(
+            [
+                self.build_up_sampling_block(in_channels=self.num_channels[d + 1], out_channels=self.num_channels[d])
+                for d in range(self.depth - 1, self.min_extract_level - 1, -1)
+            ]
+        )
+        self.decode_convs = nn.ModuleList(
+            [
+                self.build_conv_block(
+                    in_channels=(2 * self.num_channels[d] if self.concat_skip else self.num_channels[d]),
+                    out_channels=self.num_channels[d],
+                    kernel_size=3,
+                )
+                for d in range(self.depth - 1, self.min_extract_level - 1, -1)
+            ]
+        )
+
+        # extraction
+        self.output_block = self.build_output_block()
+
+    def build_up_sampling_block(
+        self,
+        in_channels: int,
+        out_channels: int,
+    ) -> nn.Module:
+        return get_deconv_block(spatial_dims=self.spatial_dims, in_channels=in_channels, out_channels=out_channels)
+
+    def build_output_block(self) -> nn.Module:
+        return RegistrationExtractionBlock(
+            spatial_dims=self.spatial_dims,
+            extract_levels=self.extract_levels,
+            num_channels=self.num_channels,
+            out_channels=self.out_channels,
+            kernel_initializer=self.out_kernel_initializer,
+            activation=self.out_activation,
+        )
+
+    def forward(self, x):
+        """
+        Args:
+            x: Tensor in shape (batch, ``in_channels``, insize_1, insize_2, [insize_3])
+
+        Returns:
+            Tensor in shape (batch, ``out_channels``, insize_1, insize_2, [insize_3]), with the same spatial size as ``x``
+        """
+        image_size = x.shape[2:]
+        skips = []  # [0, ..., depth - 1]
+        encoded = x
+        for encode_conv, encode_pool in zip(self.encode_convs, self.encode_pools):
+            skip = encode_conv(encoded)
+            encoded = encode_pool(skip)
+            skips.append(skip)
+        decoded = self.bottom_block(encoded)
+
+        outs = [decoded]
+
+        # [depth - 1, ..., min_extract_level]
+        for i, (decode_deconv, decode_conv) in enumerate(zip(self.decode_deconvs, self.decode_convs)):
+            # [depth - 1, depth - 2, ..., min_extract_level]
+            decoded = decode_deconv(decoded)
+            if self.concat_skip:
+                decoded = torch.cat([decoded, skips[-i - 1]], dim=1)
+            else:
+                decoded = decoded + skips[-i - 1]
+            decoded = decode_conv(decoded)
+            outs.append(decoded)
```

Comment:
this line is a bug, It has been fixed in https://github.com/DeepRegNet/DeepReg/blob/main/deepreg/model/backbone/u_net.py#L556

in short, the decoded should be appended in a reverse order

Reply:
The extraction block takes care of the reversion in order.

https://github.com/Project-MONAI/MONAI/blob/6979a6708caed1932655be02baa411b6caf4842c/monai/networks/blocks/regunet_block.py#L262-L268